play/src/main/scala/caliban/PlayJson.scala

```diff
@@ -0,0 +1,105 @@
+package caliban
+
+import caliban.Value._
+import caliban.parsing.adt.LocationInfo
+import play.api.http.Writeable
+import play.api.libs.json.Writes._
+import play.api.libs.json._
+
+object PlayJson {
+
+  //// JSON Reads
+  private def jsonToInputValue(json: JsValue): InputValue = json match {
+    case JsNull             => NullValue
+    case boolean: JsBoolean => BooleanValue(boolean.value)
+    case JsNumber(value) =>
+      if (value.isValidInt) IntValue(value.toInt)
+      else if (value.isValidLong) IntValue(value.toLong)
+      else if (value.isWhole) IntValue(value.toBigInt)
+      else FloatValue(value)
+    case JsString(value) => StringValue(value)
+    case JsArray(array) =>
+      InputValue.ListValue(array.toList.map(jsonToInputValue))
+    case JsObject(obj) =>
+      val builder = Predef.Map.newBuilder[String, InputValue]
+      builder.sizeHint(obj)
+      obj.foreach { case (k, v) => builder += (k -> jsonToInputValue(v)) }
+      InputValue.ObjectValue(builder.result())
+  }
+
+  implicit val readsInputValue: Reads[InputValue] =
+    json => JsSuccess(jsonToInputValue(json))
+
+  implicit val readsGraphQLRequest: Reads[GraphQLRequest] =
+    Json.reads
+
+  //// JSON Writes
+
+  // ResponseValue
+  private def responseValueToJson(response: ResponseValue): JsValue = response match {
+    case v: Value                          => valueToJson(v)
+    case ResponseValue.ListValue(values)   => JsArray(values.map(responseValueToJson))
+    case ResponseValue.ObjectValue(fields) => JsObject(fields.map { case (k, v) => k -> responseValueToJson(v) })
+    case s: ResponseValue.StreamValue      => JsString(s.toString)
+  }
+
+  private def valueToJson(value: Value): JsValue = value match {
+    case Value.NullValue     => JsNull
+    case value: IntValue     => JsNumber(BigDecimal(value.toBigInt.bigInteger))
+    case value: FloatValue   => JsNumber(value.toBigDecimal)
+    case StringValue(value)  => JsString(value)
+    case BooleanValue(value) => JsBoolean(value)
+    case EnumValue(value)    => JsString(value)
+  }
+
+  implicit val writesResponseValue: Writes[ResponseValue] = responseValueToJson _
+
+  private def handleError(err: Any): JsValue = err match {
+    case ce: CalibanError => errorValueEncoder.writes(ce)
+    case _                => Json.obj("message" -> JsString(err.toString))
+  }
+
+  // CalibanError
+  private implicit val locationToJson: Writes[LocationInfo] =
+    li => Json.obj("locations" -> Json.arr(Json.obj("line" -> li.line, "column" -> li.column)))
+
+  private def encodeLocationInfoAndMessage(li: Option[LocationInfo], message: String) =
+    li.fold(Json.obj())(li => Json.obj("locations" -> li)) ++ Json.obj("message" -> message)
```

Comment:
Can simplify to `Json.obj("message" -> message, "locations" -> li)`

Reply:
Won't that change the behavior?

Reply:
On second thought, it does change the behavior however trivial it is. I feel better leaving it behave the same with circe, wdyt?

Reply:
How does the behavior change?

Reply:
Actually why does the locationinfo writes also pass it’s key name?

Reply:
Thanks for the catch! 
I meant your simplification would add field `"locations": null`, when `LocationInfo` is empty, where currently it drops the key altogether. 
Current circe impl drops the null values.